### PR TITLE
Refactor the loading and generating of HTML.

### LINKED ISSUE #56: Power cards not loading all dropdown menus
image

Despite starting with 4 power cards, it only loads one. I suspect the reason is this - the webpage is built BEFORE the iframes are loaded... so it doesn't even know that there are 4 power cards. At the time, there's only one - the blank card built into the empty powerCards object.

### LINKED ISSUE #167: Issue moving back and forth between components since PreviewFrame was made into a component
If I move back and forth from spirit board to other components and back to spirit board (slowly), sometimes it fails to load the spirit board properly. I'm not sure why, but I think this issue started since we moved over to PreviewFrame being its own svelte thing.

### PR DESCRIPTION
**Pull request by tomprince**
This makes a number of changes to how HTML is handled:
- It uses [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/fetch) and [`DomParser`](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser) to load HTML directly (whether the initial demo, from the load button, or from the examples), rather than using a hidden `iframe` to load it. This makes it easier to then parse and reload the preview when it is complete.
- Directly generate the HTML that is passed to the preview window or saved, rather than mutating the previously loaded HTML. This avoid the need to keep around a loaded document to be able to mutate it.
- Explicitly specify the `<link rel="stylesheet">` and `<script>` tags to include in the `<head>` of the preview frame, rather than having them included as a side-effect of mutating a previously loaded document.

**Note**: This does slightly change the format that things are exported in. Specifically, it doesn't have the `<head>`, `<body>` or `<link rel="stylesheet">` tags anymore (the existing format did not have any `<html>` or `<script>` tags already). I don't think this is a significant regression, as the old files would already need to be edited to use them locally. I believe that this is forwards and backwards compatible, with respect to loading in the app (though I only tested spirit board fronts and power cards).

I think this should fix #167. I think it also fixes #56.

### CODE REVIEW

**Comment by tomprince** [comment]
It turns out I missed an issue with generating images, which I think is now fixed.

**Comment by resonant-gamedesign** [comment]
Assuming this works, this seems like a good change. 
My concern here is losing the window.onload in the functions. I think removing those will prevent the template from running. Any way you can do this without losing those? Open to ideas (ie maybe a check of some kind?)

**Comment by tomprince** [comment]
Similar to what I suggested in https://github.com/neubee/spirit-island-builder/pull/168#issuecomment-1374819662, there could be an additional script that just has that `onLoad`.

**Comment by resonant-gamedesign** [comment]
Ok, a script that the template files call that runs startmain for them.
That change would break everyone's existing template files...(I think)
because we'd need to call it in the header.

Is there another solution? One we can do on the Builder side?

File: src/routes/spirit-board/index.svelte [hunk 1]
```diff
@@ -147,211 +147,187 @@
...
+      let newRuleName = document.createElement("special-rules-subtitle");
+      newRuleName.textContent = rule.name;
+      let newRuleEffect = document.createElement("special-rule");
+      newRuleEffect.innerHTML = rule.effect;
+      specialRulesContainer.appendChild(newRuleName);
+      specialRulesContainer.appendChild(newRuleEffect);
+    });
+
+    const right = document.createElement("right");
+    board.appendChild(right);
```
Comment (resonant-gamedesign) [inline src/routes/spirit-board/index.svelte:208]: Previously, we had a blank HTML template spirit to work from, so we didn't have to add a <right> element. Now we're building from the ground up so we add everything. Is that right? Just making sure I understand. This function now builds a fresh HTML fragment that looks very similar to what used to live in 'holder-frame', right?

**Comment by resonant-gamedesign** [comment]
something like
if(!builder){ startmain }

Is there a way to figure out if we're in the Builder or not?

File: src/routes/spirit-board/index.svelte (hunk 1 above)
Comment (tomprince) [reply src/routes/spirit-board/index.svelte:208]: Yes, exactly.

**Comment by tomprince** [comment]
> That change would break everyone's existing template files because we'd need to call it in the header.

If requiring *any* changes to the files people used is out of the question, you could do something like move `general.js` to `general-common.js` in the template repository, and put
```
{
    let script = document.createElement("script");
    let baseURI = new URL(document.currentScript.src, document.baseURI);
    script.src = new URL("general-common.js", baseURI);
    document.head.append(script);
    window.addEventListener("load", () => window.startMain());
}
```
in `general.js`.

**Comment by resonant-gamedesign** [comment]
does that need to be in a function?

**Comment by resonant-gamedesign** [comment]
nevermind. it works! Thanks.
Ok so what do we need to change in the Builder js files?

**Comment by tomprince** [comment]
> nevermind. it works! Thanks. Ok so what do we need to change in the Builder js files?

Nothing, with what I suggested. If the file names need to exactly (!) match up between repos, we'd need to change the header fragments in `src/routes/*/index.svelte` to point at the `general-common.js` file instead.

**Comment by resonant-gamedesign** [comment]
Lets do that. I like to be able to copy over my changes 1 to 1.

**Comment by resonant-gamedesign** [comment]
(I don't know how to make that change in your PR, otherwise I'd do it myself, sorry)

**Comment by resonant-gamedesign** [comment]
I have it called general-common.js in the template, but I'll just update that to common.js (and then update general.js to point to common.js).

**Comment by tomprince** [comment]
@neubee Anything blocking this (and #181/#187) from landing? I can't see if there are any conflicts from my phone, but if there are, I'll rebase in the next hour or so.

**Comment by resonant-gamedesign** [comment]
I'm reviewing now. I was having issues switching between different SI
components (ie. when I switch away from Spirit Board Play Side, and then
back to it, it didn't load). Chrome, Windows.
I'm looking at it now.

**Comment by resonant-gamedesign** [comment]
Here's what I'm getting as an error when I try to switch back to a component that was previously loaded:
image

**Comment by resonant-gamedesign** [comment]
I haven't been able to solve it myself.  In my testing, I think the issue is that copyHTMLFrom doesn't actually change previewIframe. I think maybe the previewIframe's srcdoc attribute is overriding the update cause by previewIframe.contentDocument.body.replaceChildren

**Comment by tomprince** [comment]
It looks like chrome behave slightly differently in terms of timing for updating the iframe's `contentDocument` attribute. I've pushed a fix that appears to work in both firefox and chrome.

**Comment by resonant-gamedesign** [comment]
This is now working for me on Chrome. Great work Tom.
@neubee we're good to go

**Pull request merged**